### PR TITLE
feat(admin): admin update community list

### LINKED ISSUE #3261: 🚀 [Feature] GMS publish user - Frontend - Gradido communities-dialog
## 🚀 Feature

- [ ] admin interface communities dialog

see technical concept: 
https://github.com/gradido/gradido/blob/master/docu/Concepts/TechnicalRequirements/UC_publish_user_to_GMS.md#gradido-communities-dialog-gradido-efforts

### PR DESCRIPTION
**Pull request by einhornimmond**
## 🍰 Pullrequest
- Show federated communities as sub-entry from communities
- click on entry for more details
- function to add and edit GMS API Key for Home Community

### CODE REVIEW

File: backend/src/graphql/resolver/util/communities.ts [hunk 1]
```diff
@@ -63,3 +69,55 @@ export async function getCommunityByUuid(communityUuid: string): Promise<DbCommu
...
+      communityMap.set(publicKeyHex, community)
+    }
+    const community = communityMap.get(publicKeyHex)
+    if (!community?.federatedCommunities) {
+      throw new LogError('missing community after set it into map', publicKeyHex)
+    }
+    community.federatedCommunities.push(value)
+  })
+  return Array.from(communityMap.values())
+}
```
Comment (clauspeterhuebner) [inline backend/src/graphql/resolver/util/communities.ts:123]: Why simulating a RightJoin here in TypeScript when the database could select all the data like you need for you graphQL-model?

Imaging the process how both tables are filled:
1. Communities: the homeCom-Entry is created as soon the DHT starts
2. federatedCommunity: the homeCom-Entry is created as soon the DHT starts
3. federatedCommunity: the DHT writes foreignCom-Entries during federation
4. federatedCommunity: the validateCommunity-Process exchange info per direct graphql-communication and mark the foreign federatedCom as verified
5. Communities: the validateCommunity-Process starts for a verified federatedCom the authenticateCommunity-Process and creates a foreignCom-Entry in communities table
6. Communities: the authenticateCommunity-Process updates the foreignCom-Entry after successfull exchange of both communityUuids

Your master-table will be the `federatedCommunities` and for each different `publicKey` - means you need `group by publicKey` - an associated entry in `communities` could exists

File: backend/src/seeds/graphql/queries.ts [hunk 2]
```diff
@@ -156,7 +156,7 @@ export const getCommunities = gql`
       id
       foreign
       publicKey
-      url
+      endPoint
```
Comment (clauspeterhuebner) [inline backend/src/seeds/graphql/queries.ts:159]: if you cut-off the `apiVersion` from `url` and rename it now to `endpoint`, you need at least to return the apiVersion not to miss some details.

File: database/entity/0083-join_community_federated_communities/FederatedCommunity.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    name: 'updated_at',
+    type: 'datetime',
+    onUpdate: 'CURRENT_TIMESTAMP(3)',
+    nullable: true,
+  })
+  updatedAt: Date | null
+
+  @ManyToOne(() => Community, (community) => community.federatedCommunities) // Assuming you have a User entity with 'accounts' relation
+  @JoinColumn({ name: 'public_key', referencedColumnName: 'publicKey' })
+  community?: Community
```
Comment (clauspeterhuebner) [inline database/entity/0083-join_community_federated_communities/FederatedCommunity.ts:57]: Is the comment with the assumption of a user-entity with `accounts` relation valid here or simply a copy-paste-error?

**Review by clauspeterhuebner — CHANGES_REQUESTED** [review]

File: backend/src/graphql/resolver/util/communities.ts (hunk 1 above)
Comment (einhornimmond) [reply backend/src/graphql/resolver/util/communities.ts:123]: My master table is federatedCommunities
```ts
.getRepository(DbFederatedCommunity)
```
I use communityMap for sorting by publicKey

Comment (einhornimmond) [reply backend/src/graphql/resolver/util/communities.ts:123]: - I retrieve federatedCommunity entries together with community entry if exist
- for each federatedCommunity:
  - check if community already exist in communityMap by public key
  - if not create empty Community entity and insert into map 
  - push federatedCommunity into federatedCommunities array from community

Finally:
- I have a Community Entity for each different public key in federated_communities
- the contain a array with at least on federatedCommunity
- if no community entry with this public key exist in db, the community entity has only federatedCommunities
- that's reason for empty checks and copy over values from federatedCommunity in model/Community

File: backend/src/graphql/model/Community.ts [hunk 4]
```diff
@@ -1,15 +1,40 @@
...
+      const url = new URL(federatedCommunity.endPoint)
+      // use only the host part
+      this.url = url.protocol + '//' + url.host
+      this.publicKey = federatedCommunity.publicKey.toString('hex')
+      this.federatedCommunities = dbCom.federatedCommunities.map(
+        (federatedCom: DbFederatedCommunity) => new FederatedCommunity(federatedCom),
+      )
+    }
+    this.id = dbCom.id ?? 0
+    if (dbCom.foreign !== undefined) {
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/model/Community.ts:22]: the function getAllCommunities may return Communities which only contain federatedCommunities, so dbCom.foreign is undefined.

File: backend/src/graphql/model/Community.ts [hunk 5]
```diff
@@ -1,15 +1,40 @@
...
+import { FederatedCommunity } from './FederatedCommunity'
+
 @ObjectType()
 export class Community {
   constructor(dbCom: DbCommunity) {
-    this.id = dbCom.id
-    this.foreign = dbCom.foreign
+    if (dbCom.federatedCommunities && dbCom.federatedCommunities.length > 0) {
+      const federatedCommunity = dbCom.federatedCommunities[0]
+      this.foreign = federatedCommunity.foreign
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/model/Community.ts:12]: the function getAllCommunities may return Communities which only contain federatedCommunities,
so I must copy over the foreign value from a federatedCommunity

File: backend/src/graphql/model/Community.ts [hunk 6]
```diff
@@ -1,15 +1,40 @@
...
+      const url = new URL(federatedCommunity.endPoint)
+      // use only the host part
+      this.url = url.protocol + '//' + url.host
+      this.publicKey = federatedCommunity.publicKey.toString('hex')
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/model/Community.ts:16]: the function getAllCommunities may return Communities which only contain federatedCommunities,
so I must copy over the publicKey value from a federatedCommunity

File: backend/src/graphql/model/Community.ts [hunk 7]
```diff
@@ -1,15 +1,40 @@
...
 export class Community {
   constructor(dbCom: DbCommunity) {
-    this.id = dbCom.id
-    this.foreign = dbCom.foreign
+    if (dbCom.federatedCommunities && dbCom.federatedCommunities.length > 0) {
+      const federatedCommunity = dbCom.federatedCommunities[0]
+      this.foreign = federatedCommunity.foreign
+      const url = new URL(federatedCommunity.endPoint)
+      // use only the host part
+      this.url = url.protocol + '//' + url.host
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/model/Community.ts:15]: the function getAllCommunities may return Communities which only contain federatedCommunities,
so I must retrieve the url by parsing the endpoint.

File: backend/src/graphql/model/Community.ts [hunk 8]
```diff
@@ -1,15 +1,40 @@
...
+    if (dbCom.foreign !== undefined) {
+      this.foreign = dbCom.foreign
+    }
     this.name = dbCom.name
     this.description = dbCom.description
-    this.url = dbCom.url
+    this.gmsApiKey = dbCom.gmsApiKey
+    if (dbCom.url) {
+      this.url = dbCom.url
+    }
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/model/Community.ts:30]: the function getAllCommunities may return Communities which only contain federatedCommunities, so dbCom.url is undefined.

File: backend/src/graphql/model/Community.ts [hunk 9]
```diff
@@ -27,18 +52,33 @@ export class Community {
...
+  gmsApiKey: string | null
+
   @Field(() => String)
   url: string
 
+  @Field(() => String)
+  publicKey: string
+
+  @Field(() => String, { nullable: true })
+  communityUuid: string | null
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/model/Community.ts:65]: Oh.. you are right, that isn't correct. 
Maybe I was mislead by the naming. Why is it named communityUuid on the community Entity but uuid on the Community Output Model?

File: backend/src/graphql/model/Community.ts [hunk 10]
```diff
@@ -1,15 +1,40 @@
...
     this.description = dbCom.description
-    this.url = dbCom.url
+    this.gmsApiKey = dbCom.gmsApiKey
+    if (dbCom.url) {
+      this.url = dbCom.url
+    }
+    if (dbCom.publicKey && dbCom.publicKey.length === 32) {
+      this.publicKey = dbCom.publicKey.toString('hex')
+    }
+    this.communityUuid = dbCom.communityUuid
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/model/Community.ts:34]: Oh.. you are right, that isn't correct.
Maybe I was mislead by the naming. Why is it named communityUuid on the community Entity but uuid on the Community Output Model?

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 11]
```diff
@@ -164,7 +164,7 @@ describe('CommunityResolver', () => {
                 id: 3,
                 foreign: homeCom3.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[2].public),
-                url: expect.stringMatching('http://localhost/api/2_0'),
+                endPoint: expect.stringMatching('http://localhost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:167]: you right.

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 12]
```diff
@@ -186,7 +186,7 @@ describe('CommunityResolver', () => {
                 id: 1,
                 foreign: homeCom1.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[0].public),
-                url: expect.stringMatching('http://localhost/api/1_0'),
+                endPoint: expect.stringMatching('http://localhost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:189]: you right.

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 13]
```diff
@@ -175,7 +175,7 @@ describe('CommunityResolver', () => {
                 id: 2,
                 foreign: homeCom2.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[1].public),
-                url: expect.stringMatching('http://localhost/api/1_1'),
+                endPoint: expect.stringMatching('http://localhost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:178]: you right.

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 14]
```diff
@@ -236,7 +236,7 @@ describe('CommunityResolver', () => {
                 id: 3,
                 foreign: homeCom3.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[2].public),
-                url: expect.stringMatching('http://localhost/api/2_0'),
+                endPoint: expect.stringMatching('http://localhost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:239]: you right.

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 15]
```diff
@@ -247,7 +247,7 @@ describe('CommunityResolver', () => {
                 id: 2,
                 foreign: homeCom2.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[1].public),
-                url: expect.stringMatching('http://localhost/api/1_1'),
+                endPoint: expect.stringMatching('http://localhost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:250]: you right.

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 16]
```diff
@@ -258,7 +258,7 @@ describe('CommunityResolver', () => {
                 id: 1,
                 foreign: homeCom1.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[0].public),
-                url: expect.stringMatching('http://localhost/api/1_0'),
+                endPoint: expect.stringMatching('http://localhost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:261]: you right.

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 17]
```diff
@@ -269,7 +269,7 @@ describe('CommunityResolver', () => {
                 id: 6,
                 foreign: foreignCom3.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[5].public),
-                url: expect.stringMatching('http://remotehost/api/1_2'),
+                endPoint: expect.stringMatching('http://remotehost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:272]: you right.

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 18]
```diff
@@ -280,7 +280,7 @@ describe('CommunityResolver', () => {
                 id: 5,
                 foreign: foreignCom2.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[4].public),
-                url: expect.stringMatching('http://remotehost/api/1_1'),
+                endPoint: expect.stringMatching('http://remotehost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:283]: you right.

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 19]
```diff
@@ -291,7 +291,7 @@ describe('CommunityResolver', () => {
                 id: 4,
                 foreign: foreignCom1.foreign,
                 publicKey: expect.stringMatching(ed25519KeyPairStaticHex[3].public),
-                url: expect.stringMatching('http://remotehost/api/1_0'),
+                endPoint: expect.stringMatching('http://remotehost/api/'),
```
[the remark this answers is not in the document]
Comment (einhornimmond) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:294]: you right.
you don't need to comment the same thing over and over, I usually update every occurrence from that type of things.

File: backend/src/seeds/graphql/queries.ts (hunk 2 above)
Comment (einhornimmond) [reply backend/src/seeds/graphql/queries.ts:159]: right

File: database/entity/0083-join_community_federated_communities/FederatedCommunity.ts (hunk 3 above)
Comment (einhornimmond) [reply database/entity/0083-join_community_federated_communities/FederatedCommunity.ts:57]: yes a copy-paste error (._.);

File: backend/src/graphql/model/Community.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (clauspeterhuebner) [reply backend/src/graphql/model/Community.ts:12]: yes that's right, you have to transfer the `foreign`-value in any case from the entity to the graphql-transferobject, right?

File: backend/src/graphql/model/Community.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (clauspeterhuebner) [reply backend/src/graphql/model/Community.ts:16]: yes that's right, you have to transfer the `publicKey`-value in any case from the entity to the graphql-transferobject, right?

File: backend/src/graphql/model/Community.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (clauspeterhuebner) [reply backend/src/graphql/model/Community.ts:15]: yes i agree

File: admin/src/graphql/allCommunities.js [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+        id
+        apiVersion
+        endPoint
+        lastAnnouncedAt
+        verifiedAt
+        lastErrorAt
+        createdAt
+        updatedAt
+      }
+    }
```
Comment (clauspeterhuebner) [inline admin/src/graphql/allCommunities.js:27]: on my opinion causes this query the main problem of the following mismatching read/write-accesses and their mapping efforts or problems with `communities`- and `federatedCommunities`-Entities.

As you correctly recognize that in database there could exist `federatedCommunities`-Entries without a matching `communities`-entity. So how should this query here treat this constellation having no surrounding community as an anker for the associated federatedCommunity-SubList?

On my opinion the graphQL-Model have to define a GraphQL-Community-Typ containing attributes in the order of
1. with general attributs from both types of database-communities like:
- foreign
- publicKey
- url
2. with attributes only from the `federatedCommunities`-entries of this publicKey
   * several apiVersions should be read as a csv
3. then attributes only from the `communities`-entry of this publicKey

File: backend/src/graphql/model/Community.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (clauspeterhuebner) [reply backend/src/graphql/model/Community.ts:30]: please see my comment in ...graphql/allCommunities.js

File: admin/src/components/Federation/CommunityVisualizeItem.vue [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+      <b-col>
+        <div>
+          <a :href="item.url" target="_blank">{{ item.url }}</a>
+        </div>
+        <small>{{ `${item.publicKey.substring(0, 26)}…` }}</small>
+      </b-col>
+      <b-col v-b-tooltip="item.description">{{ item.name }}</b-col>
+      <b-col cols="2">{{ lastAnnouncedAt }}</b-col>
+      <b-col cols="2">{{ createdAt }}</b-col>
+    </b-row>
```
Comment (clauspeterhuebner) [inline admin/src/components/Federation/CommunityVisualizeItem.vue:14]: please see my comment in ...graphql/allCommunities.js

The ui-component to render the values for one _community_ have to be designed to treat the database constellation of existing `federatedCommunities`-entries but missing `communities`-entry and several `federatedCommunities`-entries for the same `publicKey`

**Review by clauspeterhuebner — CHANGES_REQUESTED** [review]

File: admin/src/components/Federation/CommunityVisualizeItem.vue (hunk 21 above)
Comment (einhornimmond) [reply admin/src/components/Federation/CommunityVisualizeItem.vue:14]: ChatGPT 3.5:

In einem gut strukturierten System sollten die verschiedenen Komponenten jeweils für ihre spezifischen Aufgabenbereiche verantwortlich sein:

#### Backend (Datenbankkonstellation):

- Das Backend, einschließlich der Datenbank und der entsprechenden Logik, sollte sich um die Verwaltung der Datenbankkonstellation kümmern. Dazu gehört die Speicherung und Verarbeitung von Daten gemäß den Anforderungen und Geschäftslogik der Anwendung.
- Das Backend ist dafür verantwortlich sicherzustellen, dass die Daten konsistent sind und in einer für das Frontend verwertbaren Form vorliegen.

#### GraphQL (Datenformat und Abfragen):

- GraphQL ist eine Abfragesprache und Ausführungsumgebung für APIs, die es dem Frontend ermöglicht, genau die Daten abzurufen, die es benötigt.
- GraphQL dient als Zwischenschicht zwischen dem Frontend und dem Backend und ermöglicht es dem Frontend, flexible Abfragen zu stellen, um genau die Daten zu erhalten, die es benötigt.
- GraphQL kann die Datenbankkonstellation im Backend verbergen und sicherstellen, dass das Frontend nur die benötigten Daten im erwarteten Format erhält.

#### Frontend (Darstellung und Benutzerinteraktion):

- Das Frontend ist dafür verantwortlich, die Daten, die es vom Backend über GraphQL erhält, auf angemessene Weise darzustellen und dem Benutzer eine benutzerfreundliche Benutzeroberfläche zu bieten.
- Das Frontend sollte sich nicht um die internen Details der Datenbankkonstellation oder der GraphQL-Abfragen kümmern müssen, sondern einfach die Daten im erwarteten Format erhalten und anzeigen können.

Indem jede Komponente ihre spezifischen Aufgabenbereiche hat und sich auf diese konzentriert, kann eine klare Trennung der Verantwortlichkeiten erreicht werden, was zu einer einfacheren Entwicklung, Wartung und Skalierbarkeit des Systems führt.

Comment (einhornimmond) [reply admin/src/components/Federation/CommunityVisualizeItem.vue:14]: > The ui-component to render the values for one community have to be designed to treat the database constellation of existing federatedCommunities-entries but missing communities-entry and several federatedCommunities-entries for the same publicKey

you can try it, but I found it really hard to archive this with vue2, it was much easier todo in backend.

File: admin/src/graphql/allCommunities.js (hunk 20 above)
Comment (einhornimmond) [reply admin/src/graphql/allCommunities.js:27]: Why there is not a community entry for every community?
In my opinion this strange behavior is the main cause of my trouble.

Comment (einhornimmond) [reply admin/src/graphql/allCommunities.js:27]: So ich habe jetzt ein eigenes Model angelegt namens AdminCommunityView wo ich meinen Code den ich vorher im Community Model ergänzt habe reingetan habe. Somit ist das jetzt getrennt vom Standard Community Model.
Ich weiß nicht genau was du mir mit deinem Text sagen willst. Mein Code funktioniert, wo ist jetzt genau das Problem?
Wie soll mir das lesen der apiVersions als csv von Nutzen sein? Ich kann dir nicht folgen.

**Review by clauspeterhuebner — APPROVED** [review]

**Pull request merged**